### ISSUE TITLE
Address pathways craziness

### ISSUE THREAD

**Issue by timothyfcook** [saxifrage/caac-map#61]
The pathway lines are too crazy and overlapping

We might need to change our genmap parameters so that resource bricks show up in an order more closely related to their pathways.

@chelseaerdner do you have any brilliant ideas on a way that we could change the design so that we don't have to make the bricks adjacent within pathways? 

Some possibilities:
- We could replace the center nodes with nodes that contain number (1,2,3,4) for the point in pathway associated with that bricks; then we could simplify the visual expression and reduce the lines shown as you go further in the pathway, i.e. the first 3 bricks could have heavy line connecting them, the next 3 could have transparent lines, the next 3 could have "barely there" transparent dashed-lines, the remaining could have just have the highlighting and numbered nodes without lines. Some variation on this so that the mess of lines is reduced.
- We could add some visual magic where the blocks within the pathway actually pop up and out of the grid, 3D like they get raised above and the rest of the grid loses color/focus.

![screen shot 2015-09-10 at 11 57 15 am](https://cloud.githubusercontent.com/assets/1823300/9793367/21c40046-57b3-11e5-93bd-45d74e070e1c.png)

**Comment by timothyfcook** [comment]
Oh! I remember when @whit537 and I were talking, we discussed the possibility of using the alleyways as the conduit for the lines. Maybe instead of the crazy node-to-node lines, we could change it so that the lines travel down the alleyways as the route, enter into the brick hit the node, and then leave back to the alleyway.

This is an option as well.

**Comment by timothyfcook** [comment]
@whit537 how hard would it be in genmap to have genmap "print" the blocks in groups, one pathway at a time so that bricks are always adjacent to those within their pathway?

**Comment by chadwhitacre** [comment]
First step on this will be to push the DAG comprehension down from JavaScript into Python.

**Comment by timothyfcook** [comment]
Which option are you going for?

**Comment by chadwhitacre** [comment]
Here's the [snip](https://github.com/saxifrage/cityasacampus/blob/30f808074a30b6e445a0432d93d69e78e9a0b233/map/map.js#L261-L263) from the current js where we use the DAG:

``` javascript
            for (var i=0, resource_id; resource_id=dag.names[i]; i++) {
                var v = dag.vertices[resource_id];
                for (var j=0, prev_id; prev_id = v.incomingNames[j]; j++)
```

That's the API we need to reproduce in `resources.json` → `topics.json`. Looks like we need:
- `names`—an Array of Strings, the ordered list of `resource_id`s in the pathway
- `vertices`—an Object mapping `resource_id` to an Object with an `incomingNames` attribute: an Array of Strings, the `resource_id`s of any previous items in the graph relative to the `resource_id` mapped

**Comment by chadwhitacre** [comment]
> Which option are you going for?

The option I think it makes sense for me to work on is to "have genmap 'print' the blocks in groups," because the other suggestions I'm seeing alter the design, and I think that only makes sense for @chelseaerdner to do. :-)

**Comment by chadwhitacre** [comment]
Python side of DAG push-down is done in #62.

JavaScript side of DAG push-down is done in https://github.com/saxifrage/cityasacampus/pull/210.

Now ready to use the DAG within `genmap` ... but https://github.com/saxifrage/caac-map/issues/63!

**Comment by chadwhitacre** [comment]
#63 is fixed.

**Comment by chadwhitacre** [comment]
_Now_ ready to use the DAG within `genmap`?

**Comment by chadwhitacre** [comment]
I'm seeing where a `resource_id` is ending up in `subtopic['dag']['names']` without a corresponding entry in `subtopic['resources']`. Hmm ...

Does that mean there's breakage in `_before` and/or `_after`?

**Comment by chadwhitacre** [comment]
``` python
(Pdb) resource_id
u'pittsburgh-filmmaker-and-center-for-the-arts-stained-glass'
(Pdb) pp subtopic['resources'].keys()
[u'mcg-youth-and-arts-electric-avenue',
 u'labs-at-carnegie-library-of-pittsburgh-the-labsy-awards',
 u'pittsburgh-filmmaker-and-center-for-the-arts-anime',
 u'pittsburgh-filmmaker-and-center-for-the-arts-graphic-design',
 u'mcg-youth-and-arts-digital-expressions',
 u'the-andy-warhol-museum-andy-warhol---pop-portraits-and-selfies',
 u'open-ideo-adding-visual-goodness-to-your-ideas-',
 u'carnegie-museums-of-art-and-natural-history-youth-museum-institute---exhibit-design-lab',
 u'the-andy-warhol-museum-online-weeeknd-factory',
 u'labs-at-carnegie-library-of-pittsburgh-open-lab',
 u'carnegie-museums-of-art--interact']
(Pdb) subtopic['id']
u'visual-communication'
(Pdb) subtopic['topic_id']
u'art-design'
(Pdb)
```

**Comment by chadwhitacre** [comment]
No, there's a record in the spreadsheet for `pittsburgh-filmmaker-and-center-for-the-arts-stained-glass`:

![screen shot 2015-09-18 at 2 29 08 pm](https://cloud.githubusercontent.com/assets/134455/9967619/ad5609b0-5e11-11e5-9e16-9eea7666d080.png)

**Comment by chadwhitacre** [comment]
Oooh! A subtopic mismatch? `pittsburgh-filmmaker-and-center-for-the-arts-stained-glass` is in `artisan`, but its `after_this` is `mcg-youth-and-arts-electric-avenue`, which is in `visual-communications`. The subtopic we're looking at is `visual-communications`.

**Comment by chadwhitacre** [comment]
Maybe this also explains the empty ending cells of pathways?

![screen shot 2015-09-18 at 2 34 26 pm](https://cloud.githubusercontent.com/assets/134455/9967735/670404d4-5e12-11e5-8f42-1a6c591b58e0.png)

**Comment by chadwhitacre** [comment]
Decision taken w/ @timothyfcook IRL is to write `genmap` (and `map.js`?) to ignore `after_this` (and `before_this`?) if it isn't in the same subtopic as the resource in question.

**Comment by chelseaerdner** [comment]
@timothyfcook & @whit53 , 

I just realized my name was throughout this thread (sorry sometimes I miss them) but, can I still be helpful here? If so, whats the best way to go about it?

**Comment by chadwhitacre** [comment]
Thanks for checking in, @chelseaerdner! :-)

I am plugging away on reducing pathway craziness in the process that actually builds the SVGs. I guess it's up to @timothyfcook whether he'd like to ask you to propose some design modifications to address the problem as well.

My current status (ignore circle size plz):

![screen shot 2015-09-18 at 2 51 35 pm](https://cloud.githubusercontent.com/assets/134455/9968156/de0fcc46-5e14-11e5-8c54-33a67159d837.png)
![screen shot 2015-09-18 at 2 51 42 pm](https://cloud.githubusercontent.com/assets/134455/9968155/de0ef366-5e14-11e5-8f5a-3881ca79fff1.png)
![screen shot 2015-09-18 at 2 51 48 pm](https://cloud.githubusercontent.com/assets/134455/9968154/de0d6154-5e14-11e5-900f-c958fab056d5.png)

**Comment by chelseaerdner** [comment]
Ooo, those lines definitely look more organized. Regardless, I think we are heading in the right direction @whit537!

**Comment by chadwhitacre** [comment]
:-)

**Comment by chadwhitacre** [comment]
We want planar graphs.

> The crossing number of a drawing is the number of pairs of edges that cross each other. If the graph is planar, then it is often convenient to draw it without any edge intersections; that is, in this case, a graph drawing represents a graph embedding. However, nonplanar graphs frequently arise in applications, so graph drawing algorithms must generally allow for edge crossings.

https://en.wikipedia.org/wiki/Graph_drawing#Quality_measures

Also: "[An Algorithm for the Graph Crossing Number Problem](http://ttic.uchicago.edu/~cjulia/papers/crossing-STOC.pdf)"

**Comment by timothyfcook** [comment]
Two rules we've decided upon:
1. Lines can't cross over other lines
2. Nodes cannot be mapped sequentially (can move any direction, but not the same direction two times in a row)

We will implement this by creating the map of blocks FIRST and then plotting resources onto available blocks by prioritizing BEST PATHWAY PLEASANTNESS (first) and APPROPRIATE BLOCK SIZE pegged to duration meta-data (second).

**Comment by chadwhitacre** [comment]
> During World War II, Hungarian mathematician Pál Turán was forced to work in a brick factory, pushing wagon loads of bricks from kilns to storage sites. The factory had tracks from each kiln to each storage site, and the wagons were harder to push at the points where tracks crossed each other, from which Turán was led to ask his brick factory problem: what is the minimum possible number of crossings in a drawing of a complete bipartite graph?

https://en.wikipedia.org/wiki/Crossing_number_(graph_theory)#History

O.o

**Comment by chadwhitacre** [comment]
Alright, so what's next here?

**Comment by timothyfcook** [comment]
Looks like we need to fix for #50 and #67 first!

**Comment by chadwhitacre** [comment]
Design: 

![navigation_09212015](https://cloud.githubusercontent.com/assets/11842603/9997611/2689ab8e-605b-11e5-8b46-1e8abc7b46b1.png)

**Comment by chadwhitacre** [comment]
The thing is that specifics of resources are now irrelevant at map building time. We just build a map and _then_ go back over it to assign resources to pieces.

**Comment by chadwhitacre** [comment]
[_Handbook of Graph Drawing and Visualization_](https://cs.brown.edu/~rt/gdhandbook/)

**Comment by chadwhitacre** [comment]
"In contrast, crossing minimization is extremely hard also practically."

**Comment by chadwhitacre** [comment]
Reassigning resources to pieces, using a random assignment function:

![screen shot 2015-09-25 at 3 24 51 pm](https://cloud.githubusercontent.com/assets/134455/10110159/973bfe72-6399-11e5-85e3-848a6fdf2d3e.png)

![screen shot 2015-09-25 at 3 25 58 pm](https://cloud.githubusercontent.com/assets/134455/10110200/c0e786ba-6399-11e5-8596-5a28111f19ff.png)

**Comment by chadwhitacre** [comment]
Landed #69, now integrating in https://github.com/saxifrage/cityasacampus/pull/272.

**Comment by chadwhitacre** [comment]
Ready to close, ya?
